File: apps/web/lib/api/progress.ts
import { apiRequest } from './client';
import { getToken } from './auth';

export interface ContinueLearningItem {
  topic: string;
  progress: number;
}

export interface DailyActivity {
  day: string;
  minutes: number;
  activities_count?: number;
}

export interface HeatmapDay {
  date: string;
  day: string;
  count: number;
  level: number;
}

export interface RecentActivityItem {
  id: number;
  activity_type: string;
  activity_title: string;
  xp_earned: number;
  created_at: string;
}

export interface UserProgressResponse {
  name: string;
  xp: number;
  level: number;
  streak: number;
  lessons_completed: number;
  roadmap_progress_percentage: number;
  quizzes_completed: number;
  challenges_passed: number;
  active_weak_topics_count: number;
  resolved_topics_count: number;
  continue_learning: ContinueLearningItem | null;
  weekly_activity: DailyActivity[];
  heatmap: number[][];
  heatmap_days: HeatmapDay[];
  recent_activity: RecentActivityItem[];
}

export function getUserProgress(token?: string): Promise<UserProgressResponse> {
  const authToken = token ?? getToken();
  return apiRequest<UserProgressResponse>('/users/progress', {
    method: 'GET',
    token: authToken || undefined,
  });
}

File: apps/api/app/schemas/progress.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel


class ContinueLearningItem(BaseModel):
    topic: str
    progress: int


class DailyActivity(BaseModel):
    day: str
    minutes: int
    activities_count: int = 0


class HeatmapDay(BaseModel):
    date: str
    day: str
    count: int
    level: int


class RecentActivityItem(BaseModel):
    id: int
    activity_type: str
    activity_title: str
    xp_earned: int
    created_at: datetime

    model_config = {
        "from_attributes": True
    }


class UserProgressResponse(BaseModel):
    name: str
    xp: int
    level: int
    streak: int
    lessons_completed: int = 0
    roadmap_progress_percentage: int = 0
    quizzes_completed: int = 0
    challenges_passed: int = 0
    active_weak_topics_count: int = 0
    resolved_topics_count: int = 0
    continue_learning: Optional[ContinueLearningItem] = None
    weekly_activity: List[DailyActivity] = []
    heatmap: List[List[int]] = []
    heatmap_days: List[HeatmapDay] = []
    recent_activity: List[RecentActivityItem] = []
